src/core/Oscillator.cpp

```diff
@@ -88,6 +106,238 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
 	}
 }
 
+void Oscillator::generateSineWaveTable(sample_t * table)
+{
+	// sinewaves are the only continuous waves containing no harmonics
+	// hence by definition all wavetables contain a single signal
+	// https://en.wikipedia.org/wiki/Sine_wave
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = sinf(((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH) * F_2PI);
+	}
+}
+
+void Oscillator::generateSawWaveTable(int bands, sample_t * table)
+{
+	// sawtooth wave contain both even and odd harmonics
+	// hence sinewaves are added for all bands
+	// https://en.wikipedia.org/wiki/Sawtooth_wave
+	float max = 0;
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = 0.0;
+		for (float n = 1; n <= bands; n++)
+		{
+			table[i] += ((int)n%2 ? 1.0f : -1.0f) / n * sinf(F_2PI * i * n / (float)OscillatorConstants::WAVETABLE_LENGTH);
```

Comment:
I looked through it briefly, but I think I'll need some more time with that to make sure I understand exactly what it does (so that I don't break something if I try to implement the `n+1` form `n` idea).

Precomputing some terms before the loop makes sense, I make some measurements if and how much it helps (if the compiler is smart enough it could already be doing it?)

Reply:
I played with this puzzle today and the optimization you proposed seems to work pretty good: the slowest waveform took around 2300 ms to compute in the original form (on a relatively old i3 CPU), and takes only around 130 ms after copying and re-using the table generated in the previous step.

I had to move normalization outside of the shape generating function (otherwise the re-used wave table would be normalized again and again and the older values would become distorted) but I apparently introduced a bug somewhere that causes the result to be around 0.5 dBFS lower than it should be.

After figuring that one out, I just need to rewrite the square and triangle shapes, so maybe today, maybe tomorrow.

Reply:
Cool. Leave your time, I think I'll need the whole weekend to dive into `SampleBuffer` :grin:

Reply:
Ah, right. I think I can see the issue. It's the ringing around sharp edges, as illustrated for example on this square wave:
![screen](https://user-images.githubusercontent.com/20932751/122994789-14fd2400-d3a9-11eb-8d0b-bc60b0822663.png)

The normalization is done by searching for the maximum value and then dividing all samples by the value, getting a peak value of `1.0`. And the ringing causes a "false maximum" to be found.. So the problem was already present before, I just did not notice before the difference is so small. Testing the optimization just made me look closer. :)

Reply:
Initialization timings before and after optimizations and fixes:
```
             before    after
fft took    37.9217   38.2041
sqr took    993.64    95.5435
tri took    1506.5    109.332
saw took    2346.8    162.384
```
Overall more than ten times faster, but still "slow enough" that I think it still benefits from parallelization (mainly on slow but multi-core machines like Raspberry Pi).

Reply:
These are milliseconds, correct?

Reply:
Yes. It jumps around a bit (as expected), around +- 10 ms. I will push the code tomorrow; I tried to clean it up and commit, but I kept finding new possible optimizations to try. None of them led to any further improvement, but I still need to re-test it and make sure I did not forgot any testing code in place, which is something I need to do while not falling asleep..

Reply:
The ringing is known as [Gibbs phenomenon](https://en.wikipedia.org/wiki/Gibbs_phenomenon). The magnitude of ringing at high `n` limit is mathematically known.
I also have a question: should we limit the peak amplitude or preserve the magnitude of the lowest harmonic? @curlymorphic Could you answer this question?

Reply:
I'm not sure I know the correct answer to this question and would welcome corrections but I would say the peak should be limited to -1.0 +1.0. It would be bad practice to have the audio signal outside of this range and could well lead to digit clipping in any additional processing in the chain.  A simple example of this would be any nonlinear processing that uses tanh. If the source is outside the range it would require the user to attenuate the signal as It would distort when played back, not the nice warm analog distortion but the horrid digital clipping.

If the fundamental signals are used to set the level, this would be the same as not normalizing, To keep all waveform fundamental signal the same it would require normalizing all the waveforms by the same amount, i.e applying the same attention to the sine as to the waveform with the greatest peaks,  to  keep all signals within -1 +1, resulting in a very quiet sinewave.

 Due to the harmonics adding to the perceived loudness I see no issues with the fundamental frequency being lower for a signal with more harmonics. A simple demonstration of this can be achieved by playing a normalized sine wave, then a normalized bandlimited sawtooth. If your confident that your hearing is similar in both ears you could pan the signal hard left and right and adjust the levels until the perceived loudness is the same.

Reply:
Sorry, I should have mentioned it here that I consider the normalization problem solved (I only dropped a short note in a comment I posted after the latest commit but did not go into detail).

First, I also thought that maybe normalizing by the maximum is a good thing (to avoid clipping), but there are many other, more likely causes of clipping (like simply mixing the three TripleOsc waveforms together, creating large peaks when the phase happens to align in a way that all the peaks add up) that would come into play much sooner. So I think the ringing is of no practical consequence compared to that.

Apart from the obvious inconsistency with maximum-based normalization (switching to the wavetable oscillator mode reduces volume, which is likely not expected by the user), the bigger problem is that not only the amount of ringing changes with the number of frequency bands included in the signal, but the maximum detection becomes less reliable when high frequencies are included: if the peak of the "real mathematical waveform" happens to be in the same location where a sample is available, the detected maximum will be higher than if the peak falls between two samples. So maximum-based normalization seems to be a no-go any way I look at it.

Normalizing based on the expected magnitude of ringing is an interesting idea and would solve another problem with "raw maximum-based normalization" (waveforms with high amount of frequency bands were becoming louder and louder as the ringing became "tighter" and the peak value used for normalization dropped), but it would still run into the sampling unreliability.

Anyway, it turned out the formulas used to generate the waveforms could be simply modified so that they directly produce values in the range -1.0 to 1.0 (in theory, if the sum continues to infinity). That's what I meant in the last comment when I said the normalization is now "baked in". So the peaks are now above 1.0 due to the ringing, but the magnitude of each frequency band matches the spectrum of the original, non-band-limited waveforms, preserving the loudness.

If we really want to go the way of "avoid any clipping, even if switching to wavetable oscillator versions reduces volume", we could simply go through all the generated tables for a shape, find the maximum across all of them, and use _that_ to normalize all tables for the given shape. But it feels to me as a weird heuristic rather than a nice mathematical solution.

Reply:
Looks like considerable thought has been applied to the normalization, I like the concept of matching the original loudness, I would assume this to give a better user experience than how I left the old PR.  On reflection, I feel my above post is incorrect and the current proposal is a good improvement.

@he29-net Thank you for the detailed explanation.